fil-proofs-tooling/src/bin/update_tree_r_cache/main.rs

```diff
@@ -64,27 +132,206 @@ fn run_update(sector_size: usize, cache: PathBuf, replica_path: PathBuf) -> Resu
+                build_tree_r_last::<SectorShape32GiB>(sector_size, &cache, &replica_path)?;
+            (tree_r_last.root(), base_tree_roots)
+        }
+        SECTOR_SIZE_64_GIB => {
+            let (tree_r_last, base_tree_roots) =
+                build_tree_r_last::<SectorShape64GiB>(sector_size, &cache, &replica_path)?;
+            (tree_r_last.root(), base_tree_roots)
+        }
+        _ => panic!("Unsupported sector size"),
+    };
+
+    Ok((tree_r_last_root, base_tree_roots))
+}
+
+fn run_inspect(sector_size: usize, cache: PathBuf, replica_path: PathBuf) -> Result<()> {
+    let (_tree_count, base_tree_leafs, configs, replica_config) =
+        get_tree_r_info(sector_size, &cache, &replica_path)?;
+    let tree_r_last_root =
+        get_tree_r_last_root(base_tree_leafs, sector_size, &configs, &replica_config)?;
+    let p_aux = get_persistent_aux(&cache)?;
+
+    println!("CommRLast from p_aux: {:?}", p_aux.comm_r_last);
+    println!(
+        "CommRLast [cached tree_r_last root]: {:?}",
+        tree_r_last_root
+    );
+    let status = if tree_r_last_root == p_aux.comm_r_last {
+        "MATCH"
+    } else {
+        "MIS-MATCH"
+    };
+    println!("Cached inspection shows a {} of CommRLast", status);
+
+    Ok(())
+}
+
+fn run_verify(sector_size: usize, cache: PathBuf, replica_path: PathBuf) -> Result<()> {
+    let (tree_count, base_tree_leafs, configs, replica_config) =
+        get_tree_r_info(sector_size, &cache, &replica_path)?;
+    let base_tree_len = get_merkle_tree_len(base_tree_leafs, OCT_ARITY)?;
+
+    let match_str = |a, b| -> &str {
+        if a == b {
+            "MATCH"
+        } else {
+            "MIS-MATCH"
+        }
+    };
+
+    // First, read the roots from the cached trees on disk
+    let mut cached_base_tree_roots: Vec<DefaultTreeDomain> = Vec::with_capacity(tree_count);
+    for i in 0..tree_count {
+        let store = LCStore::new_from_disk_with_reader(
+            base_tree_len,
+            OCT_ARITY,
+            &configs[i],
+            ExternalReader::new_from_config(&replica_config, i)?,
+        )?;
+        cached_base_tree_roots.push(store.last()?);
+    }
+
+    // Retrieve the tree_r_last root from the cached trees on disk.
+    let tree_r_last_root =
+        get_tree_r_last_root(base_tree_leafs, sector_size, &configs, &replica_config)?;
+
+    // Read comm_r_last from the persistent aux in the cache dir
+    let p_aux: PersistentAux<DefaultTreeDomain> = {
+        let p_aux_path = cache.join(CacheKey::PAux.to_string());
+        let p_aux_bytes = std::fs::read(&p_aux_path)
+            .with_context(|| format!("could not read file p_aux={:?}", p_aux_path))?;
+
+        deserialize(&p_aux_bytes)
+    }?;
+
+    // Rebuild each of the tree_r_last base trees (in a new temp dir so as not to interfere
+    // with any existing ones on disk) and check if the roots match what's cached on disk
+    let tmp_dir = TempDir::new("tree-r-last-verify")?;
```

Comment:
I might not completely follow the intended usage, so this question may not be quite right.

Under what circumstances would I want to go to the trouble of rebuilding the tree — in order to check the stored versions of the tree — but then discard the tree if the stored one is wrong?

If the answer is, "You wouldn't. If there is a mismatch, we will repair the existing trees by replacing with the correct ones just generated," then great. If not, perhaps we should consider that?

We should at least make sure there's no intended workflow in which a problem would be detected as above, and the solution would then involve rebuilding the tree *again* to replace.

Reply:
This is a verify command (separate from rebuild).

If the replica data is bad and the cached trees mis-match, there's no point in rebuilding at all.  

This command verifies that the replica data has not changed and that the cached trees match based on that assumption.  If there's a mis-match, we don't know which is correct, but will report that mis-match.